### ISSUE TITLE
Making the experimental branch main

### ISSUE THREAD

**Issue by orionz** [automerge/automerge-rs#279]
The experimental branch has come a long way and will soon be ready to replace to main branch.  Wanted to open this issue to discuss what should be done before making the switch.

Off the top of my head

Blocking issues: 
1. Make sure the readme is really up to snuff and points people to the old version if they are indeed looking for a backend for the js frontend.

Nonblocking Issues (but we're close so maybe we can get them in):
1. Cut an official `automerge-wasm` npm package.
2. Cut an official `automerge-rs` cargo crate.

Reasons To Delay:
1. The C API is not ready (but the failings of the old C api were central to the rewrite)
2. The `automerge-wasm` api has gotten some real use - but the `automerge-js` has not (although it passes the test suite.

I'm sure there are more.

**Comment by alexjg** [comment]
There's two main things I think we should do before we move to `main`, just so we mostly have feature parity.

* Remove the `js-sys` and `wasm-bindgen` dependencies from the core crate.
* Port the CLI crates across from `main`

I think there are a bunch more things we need to do before we do a crates.io release. But that should be a separate issue I think.

**Comment by orionz** [comment]
Perhaps we can feature flag the wasm-bindgen?  Without those I can't print debug from within rust when running in wasm.  Also I can't implement `From<AutomergeError> for JsValue` in the wasm crate because neither type is declared there and that from removes a ton of noise from the crate (allows me to do `doc.call()?` in a hundred places instead of `doc.call().map_err(automerge_to_js_err)`

**Comment by orionz** [comment]
Also - the typescript api really needed a `doc.make()` - which is just `doc.set()` but only object typers, since the whole `Option<OpID>` return type made for some messy messy typescript.  Do we want something like this in the rust api?  How ugly is having to unwrap that option in your tests?  Id we did this then it would seem we'd also need an `insert_make()` as well .. is having the option to remove an unwrap() worth two extra api calls?

Any other api tweaks we should consider before release?

**Comment by orionz** [comment]
Making a PR for moving wasm to a feature flag

**Comment by orionz** [comment]
Looking over the cli - what's there is HEAVILY tied up with frontend/backend and the change json format - im not sure if its worth porting over - vs starting over.

**Comment by alexjg** [comment]
Yeah the "port" here is more about the command line parsing stuff. The actual interaction with automerge documents should be fairly trivial I think

**Comment by alexjg** [comment]
Regarding the `make` calls, I think that would actually be a nicer API. I think for a lot of applications users will know whether they are making an object vs setting a property. i.e. their code will end up littered with comments like `SAFETY: we are creating an object so we know this is not None`.

**Comment by jeffa5** [comment]
Would the `push` function in the wasm API also need a `push_make` counterpart if we make the distinction?

**Comment by orionz** [comment]
> Would the push function in the wasm API also need a push_make counterpart if we make the distinction?

push_make, splice_make?, it could get messy, I thought I was super clever reducing the api calls down to set/insert/del but this could baloon quite fast.  Take a moment and think if we're doing this wrong?

> What was the reason for having actor be Option<usize>? Can we just create a small random one on creation and then that can be overriden? It makes the api a bit more bloated and awkward currently.

If we always autogenerate one then this will end up in the `self.ops.m.actors.cache()` and then will get added to the saved document's headers.  Over time as documents are opened and closed we could have thousands of unused actors in the header that never wrote anything.  I could make sure to only add actors that have written to the document to the headers but i figured allowing a None state for actor was simpler.  I went through a fair amount of trouble to make sure that only actors that written changes to the history end up in the cache.  Thoughts?

> Regarding the js/wasm in the core crate I think maybe the wasm crate should implement the error mapping itself so that it can be self-contained there along with any API changes for wasm/js.

If you can show me how to do this in the wasm crate I'd be thrilled.  Right now I moved it to a feature flag in a PR.  The problem is that wasm requires that I return a `Result<_,JsValue>`.  If instead it took `Result<_,Into<JsValue>>` this would be trivial and I would have no problems but alas I have no control.  When I moved the `From<AutomergeError> for JsValue>` into automerge it removed tons of visual noise from the wasm code and I was very thankful. 

I also had a few other ideas that would change internal behavior of automerge (but would also make us not 100% compatible) ... the two changes im thinking of are - only allow text to have graphmeme values for its elements (currently you can stick a ham sandwitch in the middle of a text object and it'll take it).  And disallow sets in text - inserts and deletes only - this would make the query code MUCH simpler and easier to optimize since one visible item equals one length - where as now any number of consecutive visible items can occupy the same spot (conflicts).  This maps pretty well to how people enter text and how people expect it to behave.  In fact the argument could be made for normal lists as well (if you want to do in place editing of a list element - make it an object - otherwise only insert and delete is valid)

**Comment by orionz** [comment]
On the topic of changes to the system - one that I want to see (and martin has seemed open to it too) is to make the default behavior be to refuse to apply changes that have no common ancestor.  This will make it much harder for people to do the default automerge foot-gun of making two different blank but same shape documents and then trying to merge them together.  We could make a feature flag to do this and enable it on the automerge-js side until which time martin implements it too

**Comment by ept** [comment]
> only allow text to have graphmeme values for its elements (currently you can stick a ham sandwitch in the middle of a text object and it'll take it)

I believe the rules for what constitutes a grapheme cluster can change from one version of the Unicode standard to the next (e.g. they might introduce some kind of new combining mark that adds a feature to emoji), so I'm a bit hesitant about committing to one particular interpretation of the rules.

> And disallow sets in text - inserts and deletes only

That seems reasonable, although I think it would be good to allow objects (e.g. a nested map) to appear within text, since that can be used to model some types of rich text structures.

> In fact the argument could be made for normal lists as well (if you want to do in place editing of a list element - make it an object - otherwise only insert and delete is valid)

I would like to keep the ability to overwrite list elements in-place. One example of where this might be useful is: say you have a 3D graphics app that represents points in 3D space as lists containing three numbers. Then changing the x coordinate of that vector can be done by overwriting the first list element. It does not make sense to model this as deletion and insertion, because then if two users concurrently update the x coordinate of the same point, you would end up with a four-element list, which will not make sense from the app's point of view.

> make the default behavior be to refuse to apply changes that have no common ancestor

I'm in favour of this — I've wanted to make this change in the JS implementation too.

**Comment by ept** [comment]
> A middle ground might be - only allow strings AND (at least with the low level api) - only allow splice actions.

That seems reasonable; certainly it would be good for the API to encourage the intended correct use.

>  If modifying the text (at the api level) required the string to pass through SOME implementation of a grapheme split there'd be a lot less surprise.

We could consider including a grapheme cluster splitting algorithm in the library, so that if you pass it a string, it gets split automatically. Although we should check this doesn't increase the compiled library size too much, since I think this might require some large lookup tables. There is a [proposal to standardise such an API in JS](https://github.com/tc39/proposal-intl-segmenter), but doing it in Rust would ensure consistent behaviour of Automerge across platforms.
